pkg/backend/filestate/lock.go

```diff
@@ -0,0 +1,157 @@
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package filestate
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"github.com/pulumi/pulumi/pkg/backend"
+	"github.com/pulumi/pulumi/pkg/util/logging"
+	"os"
+	"os/user"
+	"path/filepath"
+	"strings"
+	"time"
+)
+
+type lockContent struct {
+	Pid       int       `json:"pid"`
+	Username  string    `json:"username"`
+	Hostname  string    `json:"hostname"`
+	Timestamp time.Time `json:"timestamp"`
+}
+
+func newLockContent() (*lockContent, error) {
+	u, err := user.Current()
+	if err != nil {
+		return nil, err
+	}
+	hostname, err := os.Hostname()
+	if err != nil {
+		return nil, err
+	}
+	return &lockContent{
+		Pid:       os.Getpid(),
+		Username:  u.Username,
+		Hostname:  hostname,
+		Timestamp: time.Now(),
+	}, nil
+}
+
+func (l *lockContent) String() string {
+	return fmt.Sprintf("%v@%v (pid %v)", l.Username, l.Hostname, l.Pid)
+}
+
+func (b *localBackend) checkForLock(ctxt context.Context, stackRef backend.StackReference) error {
+	allFiles, err := listBucket(b.bucket, b.LockDir())
+	if err != nil {
+		return err
+	}
+
+	var lockKeys []string
+	for i := len(allFiles) - 1; i >= 0; i-- {
+		if allFiles[i].IsDir {
+			continue
+		}
+		fileKey := allFiles[i].Key
+		if strings.HasPrefix(fileKey, b.lockPrefix(stackRef.Name())) {
+			lockKeys = append(lockKeys, fileKey)
+		}
+	}
+
+	if len(lockKeys) > 0 {
+		errorString := fmt.Sprintf("the stack is current locked by %v lock(s). Either wait for the other "+
+			"process(es) to end or manually delete the lock file(s).", len(lockKeys))
+
+		for _, lock := range lockKeys {
+			content, err := b.bucket.ReadAll(ctxt, lock)
+			if err != nil {
+				return err
+			}
+			l := &lockContent{}
+			err = json.Unmarshal(content, &l)
+			if err != nil {
+				return err
+			}
+
+			// this is kind of weird but necessary because url is a string and not a url.URL
+			url := b.url + filepath.Join("/", lock)
+			errorString = errorString + fmt.Sprintf("\n  %v: created by %v", url, l.String())
+		}
+
+		return fmt.Errorf(errorString)
+	}
+	return nil
+}
+
+func (b *localBackend) checkForLockRace(stackRef backend.StackReference) error {
+	// Check the locks to make sure ONLY our lock exists. If we find multiple locks then we know
+	// we had a race condition and we should clean up and abort
+	allFiles, err := listBucket(b.bucket, b.LockDir())
+	if err != nil {
+		b.Unlock(stackRef)
+		return err
+	}
+
+	for i := len(allFiles) - 1; i >= 0; i-- {
+		file := allFiles[i].Key
+		if strings.HasPrefix(file, b.lockPrefix(stackRef.Name())) && file != b.lockPath(stackRef.Name()) {
+			b.Unlock(stackRef)
```

Comment:
super wary about the two manual unlocks in this function.  can the unlock not be moved to the caller of this?

Reply:
have seen `strings.HasPrefix(file, b.lockPrefix(stackRef.Name())` a couple of times.  Can we encapsulate this in some way as a simple question: `isLockForThisStack(file, stackRef)` or something like that?

Reply:
Got these covered.